### ISSUE TITLE
"Unable to detect the identity for the command, it could have been one of 1 possible options"

### ISSUE THREAD

**Issue by mhectorgato** [laingsimon/Dapper.MoqTests#30]
The execption occurs, with or without this setup:
`con.Setup(q => q.Execute(It.IsAny<string>(),
                It.IsAny<object>(),
                It.IsAny<IDbTransaction>(), 
                It.IsAny<int?>(), 
                It.IsAny<CommandType>()))
                .Returns(1);`
Line of code being executed:
`connection.Execute(query, new { UserId = userId, Locations = locations }, commandType: CommandType.Text, commandTimeout: CommonConstants.SqlTimeoutSeconds);`

Exception raised :
```
Unable to detect the identity for the command, it could have been one of 1 possible options.

command: 'DELETE Access WHERE UserID = @UserID AND ID IN (@Locations1)'
Parameters: `UserId = aaaee905-86ab-4ddd-ac51-1dbbd4e9d931, Locations1 = 1`
CommandType: Text

To be able to Verify the Dapper call accurately the Command and Parameters (and return type) must be unique for every invocation of a Dapper method.
```

Most of the stack track:

```
   at Dapper.MoqTests.DapperCacheInfo.GetIdentity(MockDbCommand mockDbCommand, IIdentityComparer identityComparer)
   at Dapper.MoqTests.MockDbCommand.<>c__DisplayClass3_0.<.ctor>b__0()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Lazy`1.get_Value()
   at Dapper.MoqTests.MockDbCommand.ExecuteNonQuery()
   at Dapper.SqlMapper.ExecuteCommand(IDbConnection cnn, CommandDefinition& command, Action`2 paramReader) in C:\projects\dapper\Dapper\SqlMapper.cs:line 2827
   at Dapper.SqlMapper.ExecuteImpl(IDbConnection cnn, CommandDefinition& command) in C:\projects\dapper\Dapper\SqlMapper.cs:line 570
   at Dapper.SqlMapper.Execute(IDbConnection cnn, String sql, Object param, IDbTransaction transaction, Nullable`1 commandTimeout, Nullable`1 commandType) in C:\projects\dapper\Dapper\SqlMapper.cs:line 443
```

**Comment by ghost** [comment]
Thanks for providing, I might need some more detail. For now please leave me to take a little ok and I'll update the issue with what I find

**Comment by ghost** [comment]
On an initial look, this is where the exception is stemming from.
https://github.com/laingsimon/Dapper.MoqTests/blob/3991a286bfb9fc518b1697c604d32861a5112294/Dapper.MoqTests/DapperCacheInfo.cs#L30

It requires there to be:
- exactly 1 command invocation, as it is mentioned in the message
- where the text isn't considered to match expectations, [resulting in 0 results](https://github.com/laingsimon/Dapper.MoqTests/blob/3991a286bfb9fc518b1697c604d32861a5112294/Dapper.MoqTests/DapperCacheInfo.cs#L46)
- as a result [the check is executed again](https://github.com/laingsimon/Dapper.MoqTests/blob/3991a286bfb9fc518b1697c604d32861a5112294/Dapper.MoqTests/DapperCacheInfo.cs#L52), where it doesn't need to be, still resulting no found data, as the text doesn't natch
- this ends up in the above exception being executed

In short `Is.Any<string>()` doesn't seem to be working in your case.

**Comment by laingsimon** [comment]
I've taken a look at this a bit further and  it appears to be an issue with the Setup...

You've included the following above:
```
con
  .Setup(q => q.Execute(It.IsAny<string>(), It.IsAny<object>(), It.IsAny<IDbTransaction>(), It.IsAny<int?>(), It.IsAny<CommandType>()))
  .Returns(1);
```

but it should use a nullable `CommandType`, like this:

```
con
  .Setup(q => q.Execute(It.IsAny<string>(), It.IsAny<object>(), It.IsAny<IDbTransaction>(), It.IsAny<int?>(), It.IsAny<CommandType?>()))
  .Returns(1);
```

I've created the following sample (but not added it to the project as there are others that represent this, this demonstrates the issue if you were to try it.

```
        [Test]
        public void SingleCommandMismatch()
        {
            var connectionFactory = new Mock<IDbConnectionFactory>();
            var connection = new MockDbConnection();
            var repository = new SampleRepository(connectionFactory.Object);
            connectionFactory
                .Setup(f => f.OpenConnection())
                .Returns(connection);
            connection
                .Setup(c => c.Execute(It.IsAny<string>(), It.IsAny<object>(), It.IsAny<IDbTransaction>(), It.IsAny<int?>(), It.IsAny<CommandType?>()))
                .Returns(1);

            repository.DeleteCar("ABC123");

            connection
                .Verify(c => c.Execute(It.IsAny<string>(), It.IsAny<object>(), It.IsAny<IDbTransaction>(), It.IsAny<int?>(), It.IsAny<CommandType?>()));
        }
```
This test passes, changing the verify to use a non-nullable `CommandType` emits this failure:

```
    Moq.MockException : 
    Expected invocation on the mock at least once, but was never performed: c => c.Execute(It.IsAny<string>(), It.IsAny<object>(), It.IsAny<IDbTransaction>(), It.IsAny<int?>(), (CommandType?)It.IsAny<CommandType>())
    
    Performed invocations:
    
       Mock<MockDatabase:1> (c):
    
          MockDatabase.Execute("delete from [Cars] 
    where Registration = @registration", { registration = ABC123 }, null, null, null)
```

**Comment by mhectorgato** [comment]
Sorry for the delay in responding to your message (got pulled off of that project) - I appreciate the time and research.

In another place where I was getting that exception, I updated my setup based on your response (made the CommandType nullable):
`con.Setup(q => q.Execute(It.IsAny<string>(),
                It.IsAny<object>(),
                It.IsAny<IDbTransaction>(),
                It.IsAny<int?>(),
                It.IsAny<CommandType?>()))
                .Returns(1);`

The code being mocked:
`connection.Execute(query, values, commandType: CommandType.Text, commandTimeout: CommonConstants.SqlTimeoutSeconds);`

And when my test runs I get:
`System.InvalidOperationException: 'Unable to detect the identity for the command, it could have been one of 1 possible options.

command: 'DELETE DB,SCHEMA.TABLE WHERE UserID = @UserID AND CustomerID= @Entity AND VAL IN (@Val1,@Val2)'
Parameters: `UserID = f68ee905-86ab-4ddd-ac51-1dbbd4e9d931, CustomerID = 1, Val1 = 122, Val2 = 123`
CommandType: Text`

All of the following still resulted in the exception being thrown.

I tried different Setups: 
`It.Is<CommandType?>(t => t.Value == CommandType.Text)))`
`It.Is<CommandType>(t => t == CommandType.Text)))`

Since I am not providing a transaction, I tried this as well:
`It.Is<IDbTransaction>(t => t == null),`

Based on sl-slaing's comment about the It.IsAny<string> not working, I tried:
`It.Is<string>(s => s.StartsWith("DELETE ")),` 


I will try to clone the repo and see if I can step into it to get more, and hopefully better, information.

**Comment by mhectorgato** [comment]
Looks like the issue I am running into is in the DapperIdentityComparer https://github.com/laingsimon/Dapper.MoqTests/blob/2958772ca1c9b39bacd757014cb5bb596dc04c3a/Dapper.MoqTests/DapperIdentityComparer.cs#L11 with a TSQL IN clause.

My query is like this:
`var qry = DELETE Table WHERE VAL IN @VAL`
`var vals = new List<string>() { "abc", "def" };`

So when this runs, a break point in the DapperIdentityComparer shows:
identity.sql => `DELETE Table WHERE VAL IN @VAL`
command.CommandText =>`DELETE Table WHERE VAL IN (@VAL1, @VAL2) `

Thus no matches are found. `return identity.sql == command.CommandText && ....`

FWIW -- my parameters are passed in like:
`var values = new {  VAL = vals }`

**Comment by laingsimon** [comment]
Ah... This looks like it might be Todo with the way dapper manages parameters and the `in` SQL clause.

You're correct the user SQL clause would be:

```
ExecuteAsync("DELETE FROM XX WHERE ID = @ids", new { ids = [1,2] })
```

Dapper converts this to the database SQL below:

```
DELETE FROM XX WHERE ID in (@id1, @id2)
```

The framework is comparing the executed sql to the user SQL, where there is a difference. I didn't see it before, but can see it now.

**Comment by laingsimon** [comment]
The issue is a bit more involved as it centres around the reverse engineering of the executed SQL to the original SQL. 

There would need to be either:
1. A means to convert the executed sql to the requested SQL
2. Tolerate the difference and match them more fuzzily, as is done with white space and return lines.

I'll investigate a bit further and see what can be done. 

@mhectorgato thanks for your response and helping to clarify the issue. I'll be in touch when I've identified more details on how it can be resolved.

**Comment by laingsimon** [comment]
In the mean time you can do something similar to what you've mentioned, i.e. a like/contains/starts-with match on the SQL

**Comment by laingsimon** [comment]
@mhectorgato from version 1.7.6 (dotnet core) 1.7.12 (netfx) this issue should be resolved. These packages have been released to nuget.

**Comment by mhectorgato** [comment]
Thank you for updating it. I am still getting the same error, with similar symptoms, but different details.

The mockDbCommand has: 
`DELETE xx WHERE UserID = @UserID AND ID IN (@Locations1)`

However the QueryCache.Keys has:
`DELETE xx WHERE UserID = @UserID AND ID IN @Locations`

The new DapperCommandTextHelper class is being called, however the parameterName.Success is false.
https://github.com/laingsimon/Dapper.MoqTests/blob/c300bfc59e0716fb633f84bcb5ea921a632c3cf7/Dapper.MoqTests/DapperCommandTextHelper.cs#L18

matches.Count == 1
match.Value == (@Locations1)

Looks like the regex expression is expecting a comma which is not the query string.  As a quick test, this caused Success to be true:
`((\@.+?)\d+(\,*)`

**Comment by laingsimon** [comment]
That's a really good spot, thanks for your help finding it. I'll get a fix in asap.

**Comment by laingsimon** [comment]
Thanks @mhectorgato  for identifying and helping with this issue - a new nuget package has been released (dotnet core v1.7.7 and .net fx v1.7.13)

**Comment by mhectorgato** [comment]
Thank you @laingsimon ! This worked for me.
